(docs): Added for drain node experiment (#134)
litmuschaos/litmus-docs: docs/drain-node.md

---
id: drain-node
title: Drain Node Experiment Details
sidebar_label: Drain Node
---
------

## Experiment Metadata

| Type      | Description                                  | Tested K8s Platform                                               |
| ----------| -------------------------------------------- | ------------------------------------------------------------------|
| Generic   | Drain the node where application pod is scheduled. |  GKE, Konvoy(AWS), Packet(Kubeadm), Minikube, OpenShift |

## Prerequisites

- Ensure that the Litmus Chaos Operator is running. If not, install from [here](https://github.com/litmuschaos/chaos-operator/blob/master/deploy/operator.yaml)

> - Ensure that the `drain-node` experiment resource is available in the cluster. If not, install from [here](https://hub.litmuschaos.io/charts/generic/experiments/drain-node)

Reviewer: Can you add the command also to check the respective resource

## Entry Criteria

- Application pods are healthy on the respective Nodes before chaos injection

## Exit Criteria

> - Application pods may or may not be healthy post chaos injection

Reviewer 2: Target node are in Ready state

## Details

> - This experiment drains the node where application pod is running and ensures if it is scheduled on another available node.

Reviewer 2: ```suggestion
- This experiment drains the node where application pod is running and verifies if it is scheduled on another available node.
```

> - In the end of experiment it uncordon the given node so that it can be utilised in future.

Reviewer 2: ```suggestion
- In the end of experiment it uncordons the specified node so that it can be utilised in future.
```

## Integrations

- Drain node can be effected using the chaos library: `litmus`

## Steps to Execute the Chaos Experiment

- This Chaos Experiment can be triggered by creating a ChaosEngine resource on the cluster. To understand the values to provide in a ChaosEngine specification, refer [Getting Started](getstarted.md/#prepare-chaosengine)

- Follow the steps in the sections below to prepare the ChaosEngine & execute the experiment.

### Prepare ChaosEngine

- Provide the application info in `spec.appinfo`
- Provide the auxiliary applications info (ns & labels) in `spec.auxiliaryAppInfo`
- Override the experiment tunables if desired 

#### Supported Experiment Tunables

| Variables             | Description                                                  | Type      |
| ----------------------| ------------------------------------------------------------ |---------- |

> | APP_NODE              | Name of the node too drain                                   | Mandatory |                         

Reviewer 2: ```suggestion
| APP_NODE              | Name of the node to drain                                   | Mandatory |                         
```

Reviewer 2: Lets add the optional ENV for duration of drain

#### Sample ChaosEngine Manifest

```yaml
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: nginx-chaos
  namespace: default
spec:
  chaosType: 'infra'  # It can be app/infra
  auxiliaryAppInfo: "ns1:name=percona,ns2:run=nginx"
  appinfo:
    appns: default
    applabel: 'app=nginx'
    appkind: deployment
  chaosServiceAccount: nginx-sa
  monitoring: false
  jobCleanUpPolicy: delete
  experiments:
    - name: drain-node
      spec:
        components:
           # set node name
          - name: APP_NODE
            value: 'node-1'
```

### Create the ChaosEngine Resource

- Create the ChaosEngine manifest prepared in the previous step to trigger the Chaos.

  `kubectl apply -f chaosengine.yml`

### Watch Chaos progress

> - Setting up a watch of the applications from that node are scheduling on the other nodes in the Kubernetes Cluster.

Reviewer 2: ```suggestion
- Set up a watch on the applications originally scheduled on the affected node and verify whether they are rescheduled on the other nodes in the Kubernetes Cluster.
- Also setup a watch on the nodes to verify state changes
```

>   `watch kubectl get pods --all-namespaces `

Reviewer 2: also add for nodes
```suggestion
  `watch kubectl get pods,nodes --all-namespaces `
```

- In the end of experiment it uncordons the specified node so that it can be utilised in future.

## Integrations

- Drain node can be effected using the chaos library: `litmus`

## Steps to Execute the Chaos Experiment

- This Chaos Experiment can be triggered by creating a ChaosEngine resource on the cluster. To understand the values to provide in a ChaosEngine specification, refer [Getting Started](getstarted.md/#prepare-chaosengine)

- Follow the steps in the sections below to prepare the ChaosEngine & execute the experiment.

### Prepare ChaosEngine

- Provide the application info in `spec.appinfo`
- Provide the auxiliary applications info (ns & labels) in `spec.auxiliaryAppInfo`
- Override the experiment tunables if desired 

#### Supported Experiment Tunables

| Variables             | Description                                                  | Type       | Notes           |
| ----------------------| ------------------------------------------------------------ |----------- | --------------- |
| APP_NODE              | Name of the node to drain                                    | Mandatory  |

>                  |

Reviewer: something wrong here!!
![Screenshot from 2020-01-07 15-32-04](https://user-images.githubusercontent.com/28861964/71886855-ef536f00-3162-11ea-97db-89457929322b.png)
